linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -4208,27 +4222,65 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
     } else if (mutable) {
       withDedicatedVar(RefinedType(declaredType))
     } else {
-      val refinedType = value.tpe
+      def computeRefinedType(): RefinedType = bindingName match {
+        case _ if value.tpe.isExact || declaredType == AnyType =>
+          value.tpe
+        case _: Binding.Local =>
+          value.tpe
+        case Binding.This =>
+          /* When binding to `this`, if the declared type is not `AnyType`,
+           * we are in a situation where
+           * a) we know the value must be non-null, and
+           * b) the declaredType may bring more precise information than
+           *    value.tpe.base (typically when inlining a polymorphic method
+           *    that ends up having only one target in a subclass).
+           * We can refine the type here based on that knowledge
+           */
+          val improvedBaseType =
+            if (isSubtype(value.tpe.base, declaredType)) value.tpe.base
+            else declaredType
+          val isExact = false // We catch the case value.tpe.isExact earlier
+          RefinedType(improvedBaseType, isExact, isNullable = false)
+      }
+
       value match {
         case PreTransBlock(bindingsAndStats, result) =>
           withNewLocalDef(binding.copy(value = result))(buildInner) { tresult =>
             cont(addPreTransBindings(bindingsAndStats, tresult))
           }
 
         case PreTransLocalDef(localDef) if !localDef.mutable =>
-          buildInner(localDef, cont)
+          val refinedType = computeRefinedType()
+          val newLocalDef = if (refinedType == value.tpe) {
+            localDef
+          } else {
+            /* Only adjust if the replacement if ReplaceWithThis or
+             * ReplaceWithVarRef, because other types have nothing to gain
+             * (e.g., ReplaceWithConstant) or we want to keep them unwrapped
+             * because they are examined in optimizations (notably all the
+             * types with virtualized objects).
+             */
+            localDef.replacement match {
+              case _:ReplaceWithThis | _:ReplaceWithVarRef =>
+                LocalDef(refinedType, mutable = false,
+                    ReplaceWithOtherLocalDef(localDef))
+              case _ =>
+                localDef
+            }
+          }
+          buildInner(newLocalDef, cont)
 
         case PreTransTree(literal: Literal, _) =>
-          buildInner(LocalDef(refinedType, false,
+          buildInner(LocalDef(value.tpe, false,
```

Comment:
Is it necessary to avoid the call to `computeRefinedType` here? If yes, add a comment?

Reply:
It's not necessary, just avoiding a useless call. I've added a comment.